### ISSUE TITLE
kraken orderbook frozen with WS

### ISSUE THREAD

**Issue by davidmasin** [ccxt/ccxt#6659]
Hi,

When watching Kraken orderbook using watchOrderBook real-time using the following code, it becomes frequently "frozen" (see the attached graph
![kraken](https://user-images.githubusercontent.com/41179202/77231827-a590f700-6b9d-11ea-9fee-97eb23ff4d84.png)
), returning the same value for some period of time before it gets updated. It happens for both bids and asks. I observed the same (though less frequently) for huobipro.

This does not happen with fetchOrderBook,

Could you please check?
Thanks

- OS: Ubuntu Linux
- Programming Language version: Node v12.16.1
- CCXT version: ccxt.pro@0.0.88

```javascript
'use strict';
const ccxtpro = require ('ccxt.pro');
const fs = require ('fs');
const exchangeSettings = {timeout: 3000, enableRateLimit: true, rateLimit: 1000};
const kraken = new ccxtpro.kraken(exchangeSettings);
const symbol = 'BTC/USD';
let marketsLoaded = false;
const interval=1500;

setTimeout(async function run() {
  let currentTimestamp = Date.now();

  if (!marketsLoaded) {
    await kraken.loadMarkets ();
    marketsLoaded = true;
  }

  try {
    let orderbook = await kraken.watchOrderBook (symbol);
    fs.appendFileSync('kraken.dat',`${currentTimestamp} ${orderbook.bids[0][0]} ${orderbook.asks[0][0]}\n`);
  } catch (e) {
    console.log (e.constructor.name, e.message);
  }
  let remainigInterval = interval - Date.now() + currentTimestamp;
  if (remainigInterval < 0) remainigInterval = 0;
  setTimeout(run, remainigInterval);
}, interval);
```

**Comment by frosty00** [comment]
i mean this has nothing to do with the orderbooks not updating all it means is the tip of the orderbook is not updating, but it is possible that users are submitting and cancelling orders in the middle of the orderbook.

So instead of just looking at the tip:

```
${orderbook.bids[0][0]} ${orderbook.asks[0][0]}
```

Look at the bigger picture:

```
const book = await kraken.watchOrderBook ('ETH/BTC')
const bids = book['bids']
let bidPriceSum = 0
let bidVolumeSum = 0
for (const bid of bids) {
    bidPriceSum += bid[0]
    bidVolumeSum += bid[1]
}
console.log (bidPriceSum, bidVolumeSum)
```

and you will see that always either the bids or asks are updating (and yes i know that the sum is not a good hash function)

**Comment by kroitor** [comment]
@frosty00 as far as I understand, the problem is that the tip is updated on fetchOrderBook while it remains the same on watchOrderBook.

**Comment by kroitor** [comment]
@frosty00 the only way to debug this is to properly recreate this setting and compare the tip with fetchOrderBook vs watchOrderBook over time.

**Comment by frosty00** [comment]
> @frosty00 as far as I understand, the problem is that the tip is updated on fetchOrderBook while it remains the same on watchOrderBook, leading to a discrepancy between the two versions of the tip. There may be a small difference, but still, it should not be a too big difference and it should "catch up" over time.

im running it on my machine and i see no difference except the ws is ahead of the rest by a few calls

**Comment by frosty00** [comment]
im running

```

async function runWs () {
    while (1) {
        const book = await exchange.watchOrderBook ('ETH/BTC')
        console.log ('ws', book['bids'][0][0], book['asks'][0][0])
    }
}


async function runRest () {
    while (1) {
        const book = await exchange.fetchOrderBook ('ETH/BTC')
        console.log ('rest', book['bids'][0][0], book['asks'][0][0])
    }
}

runWs ()
runRest ()
```

**Comment by frosty00** [comment]
I would close

**Comment by frosty00** [comment]
it could be to do with his recursive setTimeout implementation, tbh the graph is extremely vague

**Comment by kroitor** [comment]
@davidmasin does that snippet [posted by @frosty00 above](https://github.com/ccxt/ccxt/issues/6659#issuecomment-602120434) work as expected for you?

**Comment by kroitor** [comment]
@davidmasin this snippet can shed some light on the issue, I've added some timestamps to @frosty00's snippet, so it prints out the current time as well as the orderbook time. Kraken does not serve a timestamp or nonce for the REST fetchOrderBook, but you can retrieve it from http headers: https://github.com/ccxt/ccxt/blob/master/examples/ccxt.pro/js/watch-vs-fetch.js

**Comment by davidmasin** [comment]
Thaks for your response.

I tried to test it, investigating complete orderBook returned from watchOrderBook, not only tips.
It is not that easy to debug as the problem does not occur immediately, but only after some time (but normally within an hour, so it is still reasonable to try).

So far, the problem occured only if I did not supply the optional agument "limit" in the watchOrderBook list of arguments or if the limit was equal to default (that is, 10). Then, the watchOrderBook tip remained fixed for some period of time (although the other values kept updating), whereas fetchOrderBook values were changing in each call. Clearly, the [0][0] value in watchOrderBook was unrealistinc, as it remained much higher than the other values.

So, for me, I will use high limit for now for my purposes. In general, however, I think it should be resolved.

This is my code which I used to print orderBook tips in each call and complete orderBooks if ithe tip became frozen for 50 calls (75 seconds):

```javascript
'use strict';
const ccxtpro = require ('ccxt.pro');
const fs = require ('fs');
const exchangeSettings = {timeout: 3000, enableRateLimit: true, rateLimit: 1000};
const kraken = new ccxtpro.kraken(exchangeSettings);
const symbol = 'BTC/USD';
let marketsLoaded = false;
const interval = 1500;
let lastPrice = 0;
let counter = 0;
let limit = 10; //it seems that the problem disappears if this is changed to 100 or higher 

setTimeout(async function run() {
  let currentTimestamp = Date.now();

  if (!marketsLoaded) {
    await kraken.loadMarkets();
    marketsLoaded = true;
  }

  try {
    let orderBookPromises = [kraken.watchOrderBook (symbol, limit), kraken.fetchOrderBook (symbol, limit)];
    let orderBooks = await Promise.all(orderBookPromises);
    
    if (orderBooks[0].bids[0][0] == lastPrice) counter++;
    else counter = 0;
    
    fs.appendFileSync('krakenTips.dat',`${currentTimestamp} ${orderBooks[0].bids[0][0]} ${orderBooks[1].bids[0][0]}\n`);
    
    if (counter > 50) fs.appendFileSync('krakenComplete.dat',`${currentTimestamp}\n${JSON.stringify(orderBooks[0].bids)}\n${JSON.stringify(orderBooks[1].bids)}\n\n`);
    lastPrice = orderBooks[0].bids[0][0];
    
  } catch (e) {
    console.log (e.constructor.name, e.message);
  }
  let remainigInterval = interval - Date.now() + currentTimestamp;
  if (remainigInterval < 0) remainigInterval = 0;
  setTimeout(run, remainigInterval);
}, interval);
```

**Comment by kroitor** [comment]
@davidmasin thanks for the followup, we will investigate further and will get back to you soon.

**Comment by kroitor** [comment]
@davidmasin unfortunately, there's several mistakes in the code snippet you've posted above, in particular:

--------------

1.

```JavaScript
  if (!marketsLoaded) {
    await kraken.loadMarkets();
    marketsLoaded = true;
  }
```

This is not required, because the markets will be loaded just once anyway – as explained in the Manual here: https://github.com/ccxt/ccxt/wiki/Manual#loading-markets

--------------

2.

```JavaScript
 let orderBookPromises = [kraken.watchOrderBook (symbol, limit), kraken.fetchOrderBook (symbol, limit)];
    let orderBooks = await Promise.all(orderBookPromises);
```

This bit ruins the purpose of WS, because Promise.all will wait for both methods. While waiting for the *fetchOrderBook* – the execution of your script will be blocked and it won't be able to *watchOrderBook* in that time – thus you're not using the WS to its full capacity.

--------------

3. Looping with timers the way you did is prone to errors – we suggest to stick to a proper async/await architecture instead.

Below I've posted a snippet of code that will output the watched price to the left and the fetched price to the right, along with the corresponding timestamps.  You can compare those timestamps to see the *fetch* lagging behind the *watch* as expected.

You can run it and watch it working for some time, then, please, report if it did generate a frozenTips file or not:

```JavaScript
'use strict';

const ccxtpro = require ('./ccxt.pro')
    , log = require ('ololog').handleNodeErrors ().unlimited
    , ansi = require ('ansicolor').nice
    , fs = require ('fs')
    , exchange = new ccxtpro.kraken ({ enableRateLimit: true })
    , symbol = 'BTC/USD'
    , limit = 10
    , threshold = 100

;(async () => {

    let lastFetchPrice
    let lastFetched

    async function watchOrderBook (exchange, symbol, limit) {
        let counter = 0
        let lastWatchPrice
        while (true) {

            try {
                const orderbook = await exchange.watchOrderBook (symbol, limit)
                const price = orderbook.bids[0][0]
                counter = (price === lastWatchPrice) ? ++counter : 0
                lastWatchPrice = price;

                const now = exchange.milliseconds ()
                const watchTimestamp = new Date (now)
                const fetchTimestamp = new Date (lastFetched)

                // fs.appendFileSync ('krakenTips.dat',`${watchTimestamp} ${lastWatchPrice} ${lastFetchPrice} ${fetchTimestamp}\n`);

                if ((counter > threshold) && (lastWatchPrice !== lastFetchPrice)) {

                    log.red (
                        exchange.iso8601 (now).dim,
                        'frozen tip detected:',
                        lastWatchPrice,
                        lastFetchPrice,
                        exchange.iso8601 (lastFetched).dim
                    )

                    fs.appendFileSync ('frozenTips.dat', `${watchTimestamp} ${lastWatchPrice} ${lastFetchPrice} ${fetchTimestamp}\n`)

                } else {

                    log.green (
                        exchange.iso8601 (now).dim,
                        price,
                        lastFetchPrice,
                        exchange.iso8601 (lastFetched).dim
                    )
                }
            } catch (e) {
                log.red (e)
                throw e
            }
        }
    }

    async function fetchOrderBook (exchange, symbol, limit) {
        while (true) {
            try {
                const orderbook = await exchange.fetchOrderBook (symbol, limit)
                lastFetchPrice = orderbook.bids[0][0]
                lastFetched = new Date (exchange.last_response_headers['Date']).getTime ()
            } catch (e) {
                log.red (e)
            }
        }
    }

    await Promise.all ([
        watchOrderBook (exchange, symbol, limit),
        fetchOrderBook (exchange, symbol, limit),
    ])

}) ()
```

@davidmasin let me know if that does not help.

**Comment by davidmasin** [comment]
Hi @koitor, after running your code for two hours, I see no problem. So it was probably somehow caused by the recursive setTimeout call, as already suggested by @frosty00. That helped me a lot, regards

**Comment by mkapuza** [comment]
Hello, I am getting a similar "frozen tip" issue as @dmnode using python, my code is as follows:

```python
def main(symbols, exchange):
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    markets = exchange.markets
    group = asyncio.gather(*[run(s, exchange) for s in symbols if s in markets])
    tasks = asyncio.ensure_future(group, loop=loop)
    loop.run_until_complete(tasks)
```
```python
async def run(symbol, exchange):
  while True:
    try:
      orderbook = await exchange.watch_order_book(symbol)
      ticker = Ticker(
          exchange=exchange.id,
          symbol=symbol,
          bid=orderbook["bids"][0][0],
          ask=orderbook["asks"][0][0],
      )
      ticker.commit()
      ticker.log()
    except Exception as e:
      print(symbol, exchange.id, e)
      continue
```
After running for a bit, despite new data coming in, the top of the order book will remain the same. This tends to happen on certain exchanges (kraken, huobipro, poloniex) far more than others, which makes me think it could be a problem on their end. Often only one side will be frozen, so for instance you can get a situation where bid > ask.

I am using the latest version of ccxtpro: 0.4.22

Do you have any advice? I tried to mimic the example code as closely as possible

```python
async def main():
    exchange = ccxtpro.poloniex({'enableRateLimit': True})
    while True:
        orderbook = await exchange.watch_order_book('ETH/BTC')
        print(orderbook['asks'][0], orderbook['bids'][0])

asyncio.get_event_loop().run_until_complete(main())
```

**Comment by frosty00** [comment]
hi, we will investigate

**Comment by kroitor** [comment]
Hi @mkapuza! We will investigate asap. In the meantime, if you could post a verbose log via gist.github.com – that would help pin it down.

```Python
async def run(symbol, exchange):
  exchange.verbose = True
  while True:
    try:
      orderbook = await exchange.watch_order_book(symbol)
      print(orderbook["bids"][0], orderbook['asks'][0])
    except Exception as e:
      print(symbol, exchange.id, e)
      continue
```

**Comment by mkapuza** [comment]
Thanks for your fast response @kroitor and @frosty00 !
 
I've posted logs [here](https://gist.github.com/mkapuza/a90e517416bf3e85ea78288dc21e9541)...

One thing to note is that I am running a separate thread for each exchange, do you think this could affect things?
```python
for exchange in exchanges:
        Thread(target=main, args=(trade_symbols, exchange,)).start()
```

As for the logs, it looks like there's a connection error and things get out of wack once we reconnect. A lot of messages are received without being printed

Let me know if there's any more info I can provide

**Comment by kroitor** [comment]
> One thing to note is that I am running a separate thread for each exchange, do you think this could affect things?

I don't think it would affect the tips of the orderbooks, but we will get back you with our findings as soon as we have them.

**Comment by kroitor** [comment]
@mkapuza thanks for the log, it's very helpful, btw! We suspect that those reconnections cause frozen tips, we are now in the process of replicating it and working on a fix. Will upload it asap and will let you know.

**Comment by mkapuza** [comment]
No problem, let me know if I can help in any other way

**Comment by kroitor** [comment]
@mkapuza a quick question, do you create the exchanges on the main thread and then pass them to the child threads? Can you plz paste a complete snippet including the exchange initialization, so that we could reproduce and debug your setup on our side? Do you have a "frozen tip" with any of the ccxt.pro examples from here: https://github.com/ccxt/ccxt/tree/master/examples/ccxt.pro/py ?

**Comment by mkapuza** [comment]
I created each exchange within each child thread, but since my last message I've now replaced threads with processes and manage them with [`supervisor`](http://supervisord.org) and not had any issues. I will check back if the frozen tips recur.

I'm adding some code to production to make it easier to identify if this happens again. Do you guys suspect that when the tip is frozen, the rest of the order book updates appropriately? This would make debugging much more difficult, because the tip can be "frozen" (i.e. unchanged) for minutes without anything being wrong.

**Comment by kroitor** [comment]
> I created each exchange within each child thread, but since my last message I've now replaced threads with processes and manage them with supervisor and not had any issues. I will check back if the frozen tips recur.

Thx!

> I'm adding some code to production to make it easier to identify if this happens again. Do you guys suspect that when the tip is frozen, the rest of the order book updates appropriately? This would make debugging much more difficult, because the tip can be "frozen" (i.e. unchanged) for minutes without anything being wrong.

We suspect it's one of these causes:
- either a bug in the orderbook baseclass (which is very unlikely)
- or a bug with reconnections in bitvavo.js
- or a bug on the exchange side
- or a bug with threading + asyncio loops

However, when we try to reproduce it on our side, no frozen tips occur with BTC/EUR on Bitvavo.
So we need your code to replicate it, something we could copy and run.

Something close to this (this is a complete program):

- https://github.com/ccxt/ccxt/blob/master/examples/ccxt.pro/py/bitvavo-watch-order-book.py

```Python
import ccxtpro
from asyncio import get_event_loop


print('CCXT Pro version', ccxtpro.__version__)


async def main(loop):
    exchange = ccxtpro.bitvavo({
        'enableRateLimit': True,
        'asyncio_loop': loop,
    })
    await exchange.load_markets()
    exchange.verbose = True
    symbol = 'BTC/EUR'
    while True:
        try:
            orderbook = await exchange.watch_order_book(symbol)
            print(orderbook['nonce'], symbol, orderbook['asks'][0], orderbook['bids'][0])
        except Exception as e:
            print(type(e).__name__, str(e))
            break
    await exchange.close()


loop = get_event_loop()
loop.run_until_complete(main(loop))
```

With this example we don't see a frozen tip on v0.4.27.
Therefore the question is – do you get it stuck with the same code (with the most recent version) ?

And if it works for you as well, then can you add and paste some other code to it to make it break with a frozen tip, so that we could reproduce the same situation on our side?

**Comment by frosty00** [comment]
also checkout - https://github.com/kroitor/ccxt.pro/pull/85

it should help to solve issues when you are listening to multiple orderbooks, i've noticed some lag on my end with the current master

**Comment by mkapuza** [comment]
Ok, I've run [this gist](https://gist.github.com/mkapuza/f526069eed8b12d6939b74117e170b47) over the weekend and got a frozen tip on kraken and bitvavo. It's a hard bug to catch because afaik the only way to know for certain is if bid > ask 

I ran the script over the weekend with a bunch of exchanges and currencies to increase the probability an error occurs and reduce the necessary run time. One main issue with this is that the logs are messy, if a frozen tip does occur it's difficult to traceback because the logs are shared. 

As mentioned above I've resolved this by putting exchanges in their own process.

**Comment by frosty00** [comment]
@mkapuza it should be fixed in #85, just waiting for it to get merged. The fact that it works by putting every exchange in it's own process makes me think it is the CPU is the limiting factor.

**Comment by mkapuza** [comment]
Thanks. If it helps, I've found certain exchanges are much more likely to have the issue, so there is an exchange component at least to some extent. I have stored bid/asks in a database and found records where bid > ask for the following exchanges:
- bequant
- bitfinex 
- bittrex 
- bitvavo
- gateio
- kraken
- okcoin

But not for the following:
- binance
- binanceje 
- binanceus
- bitmex
- coinbasepro
- currencycom
- ftx

The remaining I do not have data on.
